File: android/app/src/main/java/com/vaultsync/data/repository/SyncRepositoryImpl.kt
package com.vaultsync.data.repository

import com.vaultsync.data.database.dao.ConflictDao
import com.vaultsync.data.database.dao.SyncOperationDao
import com.vaultsync.data.database.entity.ConflictEntity
import com.vaultsync.data.database.entity.SyncOperationEntity
import com.vaultsync.domain.model.Conflict
import com.vaultsync.domain.model.ConflictResolution
import com.vaultsync.domain.model.OperationStatus
import com.vaultsync.domain.model.SyncOperation
import com.vaultsync.domain.repository.SyncRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

class SyncRepositoryImpl(
    private val syncOperationDao: SyncOperationDao,
    private val conflictDao: ConflictDao
) : SyncRepository {

    override fun observePendingOperations(): Flow<List<SyncOperation>> {
        return syncOperationDao.observePending().map { list ->
            list.map { it.toDomain() }
        }
    }

    override suspend fun getPendingOperations(): List<SyncOperation> {
        return syncOperationDao.getPending().map { it.toDomain() }
    }

    override suspend fun getRunningOperations(): List<SyncOperation> {
        return syncOperationDao.getRunning().map { it.toDomain() }
    }

    override suspend fun recordOperation(operation: SyncOperation) {
        syncOperationDao.insert(SyncOperationEntity.fromDomain(operation))
    }

    override suspend fun updateOperationStatus(
        id: String,
        status: OperationStatus,
        errorMessage: String?,
        completedAt: Long?
    ) {
        syncOperationDao.updateStatus(
            id = id,
            status = status.name,
            errorMessage = errorMessage,
            completedAt = completedAt
        )
    }

    override suspend fun incrementRetryCount(id: String) {
        syncOperationDao.incrementRetry(id)
    }

    override suspend fun getConflicts(): List<Conflict> {
        return conflictDao.getUnresolved().map { it.toDomain() }
    }

    override fun observeConflicts(): Flow<List<Conflict>> {
        return conflictDao.observeUnresolved().map { list ->
            list.map { it.toDomain() }
        }
    }

    override suspend fun recordConflict(conflict: Conflict) {
        conflictDao.insert(ConflictEntity.fromDomain(conflict))
    }

    override suspend fun resolveConflict(conflictId: String, resolution: ConflictResolution) {
        conflictDao.updateResolution(conflictId, resolution.name)
    }
}

File: ios/Sources/VaultSync/Data/Repositories/SyncRepositoryImpl.swift
import Foundation

public actor LocalSyncStore {
    private var operations: [String: SyncOperation] = [:]
    private var conflicts: [String: Conflict] = [:]

    private var opContinuations: [UUID: AsyncStream<[SyncOperation]>.Continuation] = [:]
    private var conflictContinuations: [UUID: AsyncStream<[Conflict]>.Continuation] = [:]

    public init() {}

    public func observePendingOps() -> AsyncStream<[SyncOperation]> {
        let id = UUID()
        return AsyncStream { [weak self] continuation in
            continuation.onTermination = { _ in
                Task { [weak self] in
                    await self?.removeOpContinuation(id: id)
                }
            }
            Task { [weak self] in
                await self?.addOpContinuation(id: id, continuation: continuation)
            }
        }
    }

    private func addOpContinuation(id: UUID, continuation: AsyncStream<[SyncOperation]>.Continuation) {
        opContinuations[id] = continuation
        continuation.yield(getPending())
    }

    private func removeOpContinuation(id: UUID) {
        opContinuations.removeValue(forKey: id)
    }

    public func observeConflicts() -> AsyncStream<[Conflict]> {
        let id = UUID()
        return AsyncStream { [weak self] continuation in
            continuation.onTermination = { _ in
                Task { [weak self] in
                    await self?.removeConflictContinuation(id: id)
                }
            }
            Task { [weak self] in
                await self?.addConflictContinuation(id: id, continuation: continuation)
            }
        }
    }

    private func addConflictContinuation(id: UUID, continuation: AsyncStream<[Conflict]>.Continuation) {
        conflictContinuations[id] = continuation
        continuation.yield(getUnresolvedConflicts())
    }

    private func removeConflictContinuation(id: UUID) {
        conflictContinuations.removeValue(forKey: id)
    }

    private func notify() {
        let pending = getPending()
        for c in opContinuations.values {
            c.yield(pending)
        }
        let unresolved = getUnresolvedConflicts()
        for c in conflictContinuations.values {
            c.yield(unresolved)
        }
    }

    public func getPending() -> [SyncOperation] {
        Array(operations.values.filter { $0.status == .pending })
    }

    public func getRunning() -> [SyncOperation] {
        Array(operations.values.filter { $0.status == .running })
    }

    public func recordOp(_ op: SyncOperation) {
        operations[op.id] = op
        notify()
    }

    public func updateOpStatus(id: String, status: OperationStatus, errorMessage: String?, completedAt: Date?) {
        if var op = operations[id] {
            op.status = status
            op.errorMessage = errorMessage
            op.completedAt = completedAt
            operations[id] = op
            notify()
        }
    }

    public func incrementRetry(id: String) {
        if var op = operations[id] {
            op.retryCount += 1
            operations[id] = op
            notify()
        }
    }

    public func getUnresolvedConflicts() -> [Conflict] {
        Array(conflicts.values.filter { $0.resolution == nil })
    }

    public func recordConflict(_ conflict: Conflict) {
        conflicts[conflict.id] = conflict
        notify()
    }

    public func resolveConflict(id: String, resolution: ConflictResolution) {
        if var c = conflicts[id] {
            c.resolution = resolution
            conflicts[id] = c
            notify()
        }
    }
}

public final class SyncRepositoryImpl: SyncRepository, @unchecked Sendable {
    private let store: LocalSyncStore

    public init(store: LocalSyncStore = LocalSyncStore()) {
        self.store = store
    }

    public func observePendingOperations() -> AsyncStream<[SyncOperation]> {
        AsyncStream { continuation in
            Task {
                for await ops in await store.observePendingOps() {
                    continuation.yield(ops)
                }
            }
        }
    }

    public func getPendingOperations() async throws -> [SyncOperation] {
        await store.getPending()
    }

    public func getRunningOperations() async throws -> [SyncOperation] {
        await store.getRunning()
    }

    public func recordOperation(_ operation: SyncOperation) async throws {
        await store.recordOp(operation)
    }

    public func updateOperationStatus(
        id: String,
        status: OperationStatus,
        errorMessage: String?,
        completedAt: Date?
    ) async throws {
        await store.updateOpStatus(id: id, status: status, errorMessage: errorMessage, completedAt: completedAt)
    }

    public func incrementRetryCount(id: String) async throws {
        await store.incrementRetry(id: id)
    }

    public func getConflicts() async throws -> [Conflict] {
        await store.getUnresolvedConflicts()
    }

    public func observeConflicts() -> AsyncStream<[Conflict]> {
        AsyncStream { continuation in
            Task {
                for await conflicts in await store.observeConflicts() {
                    continuation.yield(conflicts)
                }
            }
        }
    }

    public func recordConflict(_ conflict: Conflict) async throws {
        await store.recordConflict(conflict)
    }

    public func resolveConflict(conflictId: String, resolution: ConflictResolution) async throws {
        await store.resolveConflict(id: conflictId, resolution: resolution)
    }
}
